File: prime/src/main/java/com/telenordigital/prime/storage/StorageInitiatedEventExecutor.java
package com.telenordigital.prime.storage;

import com.telenordigital.prime.storage.entities.PurchaseRequest;

import java.util.HashSet;
import java.util.Set;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.ThreadFactory;

public final class StorageInitiatedEventExecutor {
    private ExecutorService executor;

    private final Object monitor = new Object();

    private final Set<PurchaseRequestListener> purchaseRequestListeners;

    public StorageInitiatedEventExecutor() {
        final ThreadFactory tf = new ThreadProducer();
        this.executor = Executors.newCachedThreadPool(tf);
        this.purchaseRequestListeners = new HashSet<>();
    }

    public void addPurchaseRequestListener(final PurchaseRequestListener listener) {

        synchronized (monitor) {
            purchaseRequestListeners.add(listener);
        }
    }

    private final class ThreadProducer implements ThreadFactory {

        private final ThreadFactory tf = Executors.defaultThreadFactory();

        @Override
        public Thread newThread(final Runnable r) {
            final Thread t = tf.newThread(r);
            t.setName("FbstorageEventHandler");
            return t;
        }
    }

    public void onPurchaseRequest(final PurchaseRequest req) {
        synchronized (monitor) {
            for (final PurchaseRequestListener l : purchaseRequestListeners) {
                applyPurchaseRequestThroughExecutor(req, l);
            }
        }
    }

    private void applyPurchaseRequestThroughExecutor(
            final PurchaseRequest req,
            final PurchaseRequestListener l) {
        executor.execute(() -> l.onPurchaseRequest(req));
    }
}

File: prime/src/main/kotlin/org/ostelco/prime/storage/StorageInitiatedEventExecutor.kt
package org.ostelco.prime.storage

import org.ostelco.prime.model.PurchaseRequest
import java.util.*
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.concurrent.ThreadFactory

class StorageInitiatedEventExecutor {
    private val executor: ExecutorService

    private val monitor = Any()

    private val purchaseRequestHandlers: MutableSet<PurchaseRequestHandler>

    init {
        val tf = ThreadProducer()
        this.executor = Executors.newCachedThreadPool(tf)
        this.purchaseRequestHandlers = HashSet()
    }

    fun addPurchaseRequestHandler(handler: PurchaseRequestHandler) {

        synchronized(monitor) {
            purchaseRequestHandlers.add(handler)
        }
    }

    private inner class ThreadProducer : ThreadFactory {

        private val tf = Executors.defaultThreadFactory()

        override fun newThread(r: Runnable): Thread {
            val t = tf.newThread(r)
            t.name = "FbstorageEventHandler"
            return t
        }
    }

    fun onPurchaseRequest(req: PurchaseRequest) {
        synchronized(monitor) {
            for (handler in purchaseRequestHandlers) {
                applyPurchaseRequestThroughExecutor(req, handler)
            }
        }
    }

    private fun applyPurchaseRequestThroughExecutor(
            req: PurchaseRequest,
            handler: PurchaseRequestHandler) {
        executor.execute { handler.onPurchaseRequest(req) }
    }
}
